modules/consensus/validtransaction.go

```diff
@@ -126,6 +126,15 @@ func validStorageProofs(tx *bolt.Tx, t types.Transaction) error {
 			segmentLen = uint64(crypto.SegmentSize)
 		}
 
+		// COMPATv1.0.0
+		//
+		// One hardfork apparently wasn't enough, the adjusted hardfork code
+		// ALSO had errors in it. A new hardfork has been tentatively scheduled
+		// for block 100,000.
+		if ((build.Release == "standard" && blockHeight(tx) < 100e3) || (build.Release == "testing" && blockHeight(tx) < 10)) && segmentLen == 0 {
+			segmentLen = uint64(crypto.SegmentSize)
+		}
```

Comment:
I'm confused by this code. It seems like if `fc.FileSize%crypto.SegmentSize == 0` (and `fc.FileSize != 0`), we want `segmentLen = crypto.SegmentSize`. But this logic only gets triggered below the hardfork height. Shouldn't it be the other way around?

I think a good convention would be to thoroughly document the behavior of the code before the hardfork. 
We may also want to use consts for the hardfork heights.

```go
var hardforkHeightSegment1, hardforkHeightSegment2 = func() (types.BlockHeight, types.BlockHeight) {
	switch build.Release {
	case "standard":
		return 21e3, 100e3
	case "testing":
		return 10, 10
	}
}()
```

```go
segmentLen := crypto.SegmentSize
if segmentIndex == leaves-1 && fc.FileSize%crypto.SegmentSize != 0 {
	segmentLen = fc.FileSize % crypto.SegmentSize
}

if blockHeight(tx) < hardforkHeightSegment1 {
	// HARDFORK 21,000
	//
	// The original behavior was to use the entire segment for a storage
	// proof. This is unacceptable because the last segment may not
	// contain crypto.SegmentSize bytes of data, but will be padded with
	// zeros because sp.Segment is an array. This means that a segment
	// ending in one zero is indistinguishable from a segment ending in 64
	// zeros.
	segmentLen = crypto.SegmentSize
} else if blockHeight(tx) < hardforkHeightSegment2 {
	// HARDFORK 100,000
	//
	// A bug was introduced when fixing the previous hardfork: it did not
	// account for the case where fc.FileSize is an even multiple of
	// crypto.SegmentSize. In this case, segmentLen should be
	// crypto.SegmentSize, but was erroneously set to zero.
	segmentLen = fc.FileSize % crypto.SegmentSize
}
```